### PR TITLE
Retain valid per-account route cache entries

### PR DESCRIPTION
**Pull request by perfloop-agent**
## Summary

When an entry in a non-account-bound route cache has an outdated account generation, remove and rebuild only that entry. Other account-and-subject entries retain their own generation checks instead of being discarded with the whole map.

The mixed-account route workload changed from **7,743 ns/op to 6,373 ns/op** (**17.7% lower**, 10 samples per revision, p<0.001).

## Why

A non-account-bound route cache stores results by account and subject. Previously, refreshing a stale A entry cleared all results, so subsequent B messages redid matching despite B's entry still being current. Deleting only the stale key refreshes A while allowing valid B entries to remain reusable.

## Measurements

The mixed workload held its work constant at 18 routed messages per operation. After an A refresh, B L1 misses and B sublist matches fell from 8/op to 0/op, while B L1 hits rose from 8/op to 16/op. The A side stayed at one miss, one hit, and one sublist match per operation. These counters show that the timed B replays reused their valid cache entries rather than rematching them.

A separate capacity-saturated cold-key workload held its work constant at one routed message per operation. Retaining the unrelated entry left two cache entries before the new cold B lookup instead of one; that lookup then performed one capacity prune instead of none, while still performing one B sublist match. Its timing was statistically unchanged at 829.55 ns/op before and 823.3 ns/op after the change (10 samples per revision, p=0.528849).

Supporting allocation measurements did not regress: the mixed workload changed from 2,720 to 2,448 B/op and 53 to 37 allocs/op; the cold-key workload changed from 168 to 144 B/op and 4 to 3 allocs/op.

## Reproduce

In a clean checkout of each revision, run the repository-native benchmarks with 10 samples:

```sh
go test ./server -run '^$' -bench '^(BenchmarkPerAccountCacheMixedAccountGeneration|BenchmarkPerAccountCacheColdKeyAfterGenerationRefresh)$' -benchtime=2s -benchmem -count=10 -cpu=1
```

The recorded comparison used baseline `a0f553bbcf1977df9be7855950dc71cdda1ae540` and revision `e522588bb7c7a5177fe57ef3e542c956b0c74945` on Linux/x86_64.

| Revision | Mixed-account median ns/op | Samples | Result |
| --- | ---: | ---: | --- |
| `a0f553bbcf1977df9be7855950dc71cdda1ae540` | 7,743 | 10 | baseline |
| `e522588bb7c7a5177fe57ef3e542c956b0c74945` | 6,373 | 10 | 17.7% lower; p<0.001 |

## Validation

The focused `TestPerAccountCacheGenerationRefresh` mechanism check passed. It observed 8 B sublist matches and 1 A sublist match before the change, versus 0 B sublist matches and 1 A sublist match after it.

```sh
go test ./server -run '^(TestNoRaceRouteCache|TestRoutePoolPerAccountStreamImport|TestConfigReloadRouteImportPermissionsWithAccounts|TestRoutePerAccountCacheDropsRemovedAccountAfterOtherAccountRefresh)$' -count=1 -v
```

`TestRoutePerAccountCacheDropsRemovedAccountAfterOtherAccountRefresh` primes A and B entries on a non-account-bound route, suppresses A-to-B route-control writes, reloads A without B, refreshes A, and then sends a B-to-A public routed message. It verifies that the removed B account neither receives that message nor remains in the route cache.

## Limits

This is a serial, in-process route-handler measurement. It does not measure end-to-end network latency, throughput under concurrency, or production account-reload frequency.

Full measurement and check evidence: https://app.perfloop.ai/t/oss/case_1bk24crq7f

---
This is an automated, human-reviewed Perfloop contribution opened by perfloop-agent. The body links to reproducible proof for the change. The change was benchmarked on perfloop/nats-server, an automation fork of nats-io/nats-server; the commit on this PR's head branch carries the proof.

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -6339,7 +6339,7 @@ func (c *client) getAccAndResultFromCache() (*Account, *SublistResult) {
 
 		if genid := atomic.LoadUint64(&sl.genid); genid != pac.genid {
 			ok = false
-			clear(c.in.pacache)
+			delete(c.in.pacache, string(c.pa.pacache))
```
[the remark this answers is not in the document]
Comment (perfloop-agent) [reply server/client.go:6342]: Good question — retained entries for other accounts self-validate against their own account's sublist generation on next lookup, and a removed account's entries are dropped on the next per-account refresh rather than retained. The two-server regression test added in this PR (`TestRoutePerAccountCacheDropsRemovedAccountAfterOtherAccountRefresh` in `server/routes_test.go`) exercises exactly the removed-account-after-other-account-refresh path. Happy to extend coverage if there's a scenario this misses.

**Comment by neilalexander** [comment]
@claude Please verify the comments of both robots

**Comment by perfloop-agent** [comment]
Thanks @neilalexander — and the review bot is right to flag the gap in my earlier claim. To be precise: a removed account's entry is only guaranteed to drop on the next per-account refresh **if that account had at least one local subscriber** (whose disconnect bumps the sublist `genid`, which is what my test exercises). For a removed account with **no local subscribers and no service imports**, nothing touches its sublist, `genid` is unchanged, and the stale entry does persist until it's size-pruned or the connection closes — my "dropped on the next refresh" wording overstated that case.

The reason it's still safe, as the review notes: that no-subscriber case is exactly the case where the cached sublist result is empty, so `len(r.psubs)+len(r.qsubs) > 0` is false and `processMsgResults` is never called — no message is delivered to a detached account. The only residue is a stats increment on the removed `Account` object, which is observable but not a correctness or safety issue.

I'll push a second test covering the no-subscriber path (removal leaves `genid` unchanged → entry retained → asserts no delivery) so the PR's coverage matches the claim rather than overshooting it. The one-line eviction change itself is unaffected.

**Comment by perfloop-agent** [comment]
Pushed the follow-up test in 809a687098bd2520e26d7ca22af5e7c0c3cbfd81 so the coverage matches the claim instead of overshooting it.

`TestRoutePerAccountCacheRetainsRemovedNoSubscriberAccountWithoutDelivery` is the sibling of the existing drop test and exercises exactly the path the review flagged: a removed account with **no local subscribers and no service imports**, where nothing bumps the sublist `genid`, so the cached entry passes the genid check and is **retained** rather than evicted.

Getting the account to genuinely leave `genid` untouched on removal took one non-obvious detail worth recording: with the default (implicit) system account, every account carries four `$SYS.REQ.*` service-import subscriptions, and `removeAllServiceImportSubs()` on teardown removes them and bumps `genid` four times — which would have made the entry stale and put us right back on the covered drop path. The test sets `no_system_account: true` so B truly has an empty sublist, which is the precise "no subscribers and no service imports" edge case from the discussion.

Flow: prime B's entry, hold A→B route control traffic (`dropRouteWritesConn`) so B keeps forwarding `b`, drop A's only local `b` subscriber so the next forwarded message re-primes the entry empty at the current generation, then reload B out of the catalog. After removal the test drives another routed `b` and asserts:

- the entry is **retained** (`route.in.pacache["B b"]` unchanged) and its `genid` is unchanged — removal touched nothing;
- the cached result is **empty** (`len(psubs)+len(qsubs) == 0`), so `route.go`'s `len(r.psubs)+len(r.qsubs) > 0` guard is false and `processMsgResults` is never reached — no delivery to the detached `Account`;
- the message is still accounted for (the routed `inMsgs` stat increments), confirming the only residue is the benign stats bump you both described;
- the account A entry sharing the route is untouched and still delivers, i.e. the eviction stays surgical.

So the review's conclusion holds under test: the retained no-subscriber entry is benign — a stats increment, never a delivery. `go test ./server -run TestRoutePerAccountCache -count=1` is green (both tests). The one-line eviction change is unaffected.

**Comment by tsenart** [comment]
Hi! Checking in to know if there's anything else I can push on Perfloop's side here. Thanks!

**Comment by perfloop-agent** [comment]
Closing this in favor of a better optimization, so maintainer attention can be spent on the stronger change.
